### PR TITLE
Enhance saas e2e tests

### PR DESCRIPTION
**Pull request by GuyAfik**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [link to the issue](https://jira-dc.paloaltonetworks.com/browse/CIAC-9425)

## Description
* Change the EDL e2e-test to validate only that 1.1.1.1, 2.2.2.2 and 3.3.3.3 indicators are in the EDL and make retries in case of exceptions
* Update the slack-ask to do try and run the playbook each with retry mechanism in case it fails

### CODE REVIEW

File: Tests/tests_e2e/content/xsoar_saas/test_e2e_xsoar_saas.py [hunk 1]
```diff
@@ -54,21 +54,29 @@ def test_edl_returns_indicators(request: SubRequest, xsoar_saas_client: XsoarSaa
...
+                except Exception as error:
+                    logging.error(f"test_edl_returns_indicators - time {i} - error\n{error}")
+
+                assert edl_response.status_code == 200, f"status code is {edl_response.status_code} " \
+                                                        f"from EDL in URL {edl_response.request.url}"
+                text = edl_response.text
+                assert "1.1.1.1" in text, f"1.1.1.1 indicator from URL {edl_response.request.url} is not in {text}"
+                assert "2.2.2.2" in text, f"2.2.2.2 indicator from URL {edl_response.request.url} is not in {text}"
+                assert "3.3.3.3" in text, f"3.3.3.3 indicator from URL {edl_response.request.url} is not in {text}"
+                break
```
Comment (MosheEichler) [inline Tests/tests_e2e/content/xsoar_saas/test_e2e_xsoar_saas.py:79]:
```suggestion
                try:
                    edl_response = xsoar_saas_client.do_long_running_instance_request(
                        instance_name, username=username, password=password
                    )
                    assert edl_response.status_code == 200, f"status code is {edl_response.status_code} " \
                                                        f"from EDL in URL {edl_response.request.url}"
                    text = edl_response.text
                    assert "1.1.1.1" in text, f"1.1.1.1 indicator from URL {edl_response.request.url} is not in {text}"
                    assert "2.2.2.2" in text, f"2.2.2.2 indicator from URL {edl_response.request.url} is not in {text}"
                    assert "3.3.3.3" in text, f"3.3.3.3 indicator from URL {edl_response.request.url} is not in {text}"
                    break

                except Exception as error:
                    logging.error(f"test_edl_returns_indicators - time {i} - error\n{error}")

```

**Review by MosheEichler — APPROVED** [review]
Nice
Please see my comment

File: Tests/tests_e2e/content/xsoar_saas/test_e2e_xsoar_saas.py [hunk 2]
```diff
@@ -54,21 +54,28 @@ def test_edl_returns_indicators(request: SubRequest, xsoar_saas_client: XsoarSaa
...
+                    )
+                    assert edl_response.status_code == 200, f"status code is {edl_response.status_code} " \
+                        f"from EDL in URL {edl_response.request.url}"
+                    text = edl_response.text
+                    assert "1.1.1.1" in text, f"1.1.1.1 indicator from URL {edl_response.request.url} is not in {text}"
+                    assert "2.2.2.2" in text, f"2.2.2.2 indicator from URL {edl_response.request.url} is not in {text}"
+                    assert "3.3.3.3" in text, f"3.3.3.3 indicator from URL {edl_response.request.url} is not in {text}"
+                    break
+                except Exception as error:
+                    logging.error(f"test_edl_returns_indicators - time {i} - error\n{error}")
```
Comment (thefrieddan1) [inline Tests/tests_e2e/content/xsoar_saas/test_e2e_xsoar_saas.py:78]: instead of time  i would use try / try #

**Review by thefrieddan1 — APPROVED** [review]

File: Tests/tests_e2e/content/xsoar_saas/test_e2e_xsoar_saas.py [hunk 3]
```diff
@@ -54,21 +54,28 @@ def test_edl_returns_indicators(request: SubRequest, xsoar_saas_client: XsoarSaa
...
-        ) as edl_instance_response:
-            instance_name = edl_instance_response.get("name")
-
-            edl_response = xsoar_saas_client.do_long_running_instance_request(
-                instance_name, username=username, password=password
-            )
-            assert edl_response.text, f'could not get indicators from url={edl_response.request.url} from ' \
-                f'instance {instance_name} with available indicators={available_indicators},' \
-                f' status code={edl_response.status_code}, response={edl_response.text}'
+        for i in range(1, 6):
```
Comment (thefrieddan1) [inline Tests/tests_e2e/content/xsoar_saas/test_e2e_xsoar_saas.py:57]: what happens if we have more than 6 failed attempts ? 
Do you fail the test ?

File: Tests/tests_e2e/content/xsoar_saas/test_e2e_xsoar_saas.py [hunk 4]
```diff
@@ -176,15 +183,27 @@ def test_slack_ask(request: SubRequest, xsoar_saas_client: XsoarSaasClient):
...
+                        # make sure the context is populated with thread id(s) from slack ask
+                        assert context.get("Slack", {}).get("Thread"), f'thread IDs do not exist in context {context}'
+                        break
+                except Exception as error:
+                    logging.error(f"test_slack_ask - time {i} - error\n{error}")
+                    # print incident raw response to get info about it in case of failures
+                    logging.debug(
+                        f'test_slack_ask incident metadata:\n{xsoar_saas_client.search_incidents(incident_response.id)}'
+                    )
+                    if i == 5:
```
Comment (thefrieddan1) [inline Tests/tests_e2e/content/xsoar_saas/test_e2e_xsoar_saas.py:205]: I think this if == 5 statement should be added to the EDL test as well

File: Tests/tests_e2e/content/xsoar_saas/test_e2e_xsoar_saas.py (hunk 3 above)
Comment (GuyAfik) [reply Tests/tests_e2e/content/xsoar_saas/test_e2e_xsoar_saas.py:57]: you are right, fixed

File: Tests/tests_e2e/content/xsoar_saas/test_e2e_xsoar_saas.py (hunk 4 above)
Comment (GuyAfik) [reply Tests/tests_e2e/content/xsoar_saas/test_e2e_xsoar_saas.py:207]: nice catch, fixed it

File: Tests/tests_e2e/content/xsoar_saas/test_e2e_xsoar_saas.py (hunk 2 above)
Comment (GuyAfik) [reply Tests/tests_e2e/content/xsoar_saas/test_e2e_xsoar_saas.py:78]: i like `time` more, makes more sense IMO

**Pull request merged**